src/typedarray.rs

```diff
@@ -0,0 +1,314 @@
+            Ok(TypedArray {
+                object: guard,
+                computed: None
+            })
+        }
+    }
+
+    fn data(&mut self) -> (*mut T::Element, u32) {
+        if let Some(data) = self.computed {
+            return data;
+        }
+
+        let data = unsafe {
+            T::length_and_data(*self.object)
+        };
+        self.computed = Some(data);
+        data
+    }
+
+    /// # Unsafety
+    ///
+    /// The returned slice can be invalidated if the underlying typed array
+    /// is neutered.
+    pub unsafe fn as_slice(&mut self) -> &[T::Element] {
+        let (pointer, length) = self.data();
+        slice::from_raw_parts(pointer as *const T::Element, length as usize)
+    }
+
+    /// # Unsafety
+    ///
+    /// The returned slice can be invalidated if the underlying typed array
+    /// is neutered.
+    ///
+    /// The underlying `JSObject` can be aliased, which can lead to
+    /// Undefined Behavior due to mutable aliasing.
+    pub unsafe fn as_mut_slice(&mut self) -> &mut [T::Element] {
+        let (pointer, length) = self.data();
+        slice::from_raw_parts_mut(pointer, length as usize)
+    }
+}
+
+impl<'a, T: TypedArrayElementCreator + TypedArrayElement> TypedArray<'a, T> {
+    /// Create a new JS typed array, optionally providing initial data that will
+    /// be copied into the newly-allocated buffer. Returns the new JS reflector.
+    pub unsafe fn create(cx: *mut JSContext,
+                         length: u32,
+                         data: Option<&[T::Element]>,
+                         result: MutableHandleObject)
+                         -> Result<(), ()> {
+        result.set(T::create_new(cx, length));
+        if result.get().is_null() {
+            return Err(());
+        }
+
+        if let Some(data) = data {
+            assert!(data.len() <= length as usize);
+            let buf = T::get_data(result.get());
+            ptr::copy_nonoverlapping(data.as_ptr(), buf, data.len());
+        }
+
+        Ok(())
+    }
+}
+
+/// Internal trait used to associate an element type with an underlying representation
+/// and various functions required to manipulate typed arrays of that element type.
+pub trait TypedArrayElement {
+    /// Underlying primitive representation of this element type.
+    type Element;
+    /// Unwrap a typed array JS reflector for this element type.
+    unsafe fn unwrap_array(obj: *mut JSObject) -> *mut JSObject;
+    /// Retrieve the length and data of a typed array's buffer for this element type.
+    unsafe fn length_and_data(obj: *mut JSObject) -> (*mut Self::Element, u32);
+}
+
+/// Internal trait for creating new typed arrays.
+pub trait TypedArrayElementCreator: TypedArrayElement {
+    /// Create a new typed array.
+    unsafe fn create_new(cx: *mut JSContext, length: u32) -> *mut JSObject;
+    /// Get the data.
+    unsafe fn get_data(obj: *mut JSObject) -> *mut Self::Element;
+}
+
+macro_rules! typed_array_element {
```

Comment:
I propose one macro branch that accepts $t, $element, $unwrap, and $length_and_data, which declares the struct and implements TypedArrayElement, and another branch that accepts all the existing arguments, invokes the previous branch, and implements TypedArrayElementCreator.